File: hashMap.js
export default class HashMap {
  constructor(capacity = 16, loadFactor = 0.75) {
    this.loadFactor = loadFactor;
    this.capacity = capacity;
    this.table = new Array(capacity);
    this.size = 0;
  }

  hash(key) {
    let hashCode = 0;

    const primerNumber = 31;
    for (let i = 0; i < key.length; i++) {
      hashCode = (primerNumber * hashCode + key.charCodeAt(i)) % this.capacity;
    }

    return hashCode;
  }

  resize() {
    const oldTable = this.table;
    this.capacity *= 2;
    this.table = new Array(this.capacity);
    this.size = 0;

    for (const bucket of oldTable) {
      if (bucket) {
        for (const [k, v] of bucket) {
          this.set(k, v);
        }
      }
    }
  }

  set(key, value) {
    const index = this.hash(key);
    if (!this.table[index]) {
      this.table[index] = [];
    }

    for (let pair of this.table[index]) {
      if (pair[0] === key) {
        pair[1] = value;
        return;
      }
    }

    this.table[index].push([key, value]);
    this.size++;
    if (this.size / this.capacity > this.loadFactor) {
      this.resize();
    }
  }

  get(key) {
    const index = this.hash(key);
    const bucket = this.table[index];
    if (bucket) {
      for (let [k, v] of bucket) {
        if (k === key) return v;
      }
    }
    return null;
  }

  has(key) {
    const index = this.hash(key);
    const bucket = this.table[index];
    if (bucket) {
      for (let [k, _] of bucket) {
        if (k === key) return true;
      }
    }
    return false;
  }

  remove(key) {
    const index = this.hash(key);
    const bucket = this.table[index];
    if (bucket) {
      for (let i = 0; i < bucket.length; i++) {
        const [k, _] = bucket[i];

        if (k === key) {
          bucket.splice(i, 1);
          this.size--;
          return true;
        }
      }
    }
    return false;
  }

  clear() {
    this.table = new Array(this.capacity);
    this.size = 0;
  }

  keys() {
    const result = [];
    for (const bucket of this.table) {
      if (bucket) {
        for (const [k, _] of bucket) {
          result.push(k);
        }
      }
    }
    return result;
  }

  values() {
    const result = [];
    for (const bucket of this.table) {
      if (bucket) {
        for (const [_, v] of bucket) {
          result.push(v);
        }
      }
    }
    return result;
  }

  entries() {
    const result = [];
    for (const bucket of this.table) {
      if (bucket) {
        for (const pair of bucket) {
          result.push(pair);
        }
      }
    }
    return result;
  }
}

File: hashmap.py
class HashMap:
    def __init__(self, capacity=16, load_factor=0.75) -> None:
        self.load_factor = load_factor
        self.capacity = capacity
        self.table = [None] * capacity
        self.size = 0

    def hash(self, key):
        hash_code = 0
        prime = 31
        for char in key:
            hash_code = (prime * hash_code + ord(char)) % self.capacity
        return hash_code

    def resize(self):
        old_table = self.table
        self.capacity *= 2
        self.table = [None] * self.capacity
        self.size = 0

        for bucket in old_table:
            if bucket:
                for k, v in bucket:
                    self.set(k, v)

    def set(self, key, value):
        index = self.hash(key)
        if self.table[index] is None:
            self.table[index] = []

        for pair in self.table[index]:
            if pair[0] == key:
                pair[1] = value
                return

        self.table[index].append([key, value])
        self.size += 1

        if self.size / self.capacity > self.load_factor:
            self.resize()

    def get(self, key):
        index = self.hash(key)
        bucket = self.table[index]
        if bucket:
            for k, v in bucket:
                if k == key:
                    return v
        return None

    def has(self, key):
        index = self.hash(key)
        bucket = self.table[index]
        if bucket:
            for k, _ in bucket:
                if k == key:
                    return True
        return False

    def remove(self, key):
        index = self.hash(key)
        bucket = self.table[index]
        if bucket:
            for i, (k, _) in enumerate(bucket):
                if k == key:
                    del bucket[i]
                    self.size -= 1
                    return True
        return False

    def clear(self):
        self.table = [None] * self.capacity
        self.size = 0

    def keys(self):
        result = []
        for bucket in self.table:
            if bucket:
                for k, _ in bucket:
                    result.append(k)
        return result

    def values(self):
        result = []
        for bucket in self.table:
            if bucket:
                for _, v in bucket:
                    result.append(v)
        return result

    def entries(self):
        result = []
        for bucket in self.table:
            if bucket:
                for pair in bucket:
                    result.append(pair)
        return result
